connector/src/main/java/org/geysermc/connector/network/translators/inventory/updater/AnvilInventoryUpdater.java

```diff
@@ -0,0 +1,461 @@
+
+                        Tag javaEnchLvl = enchantTag.get("lvl");
+                        if (!(javaEnchLvl instanceof ShortTag || javaEnchLvl instanceof IntTag))
+                            continue;
+
+                        enchantments.put(enchantment, ((Number) javaEnchLvl.getValue()).intValue());
+                    }
+                }
+            }
+        }
+        return enchantments;
+    }
+
+    private boolean isEnchantedBook(GeyserSession session, GeyserItemStack itemStack) {
+        return itemStack.getJavaId() == session.getItemMappings().getStoredItems().enchantedBook().getJavaId();
+    }
+
+    private boolean isCombining(GeyserSession session, GeyserItemStack input, GeyserItemStack material) {
+        return isEnchantedBook(session, material) || (input.getJavaId() == material.getJavaId() && hasDurability(session, input));
+    }
+
+    private boolean isRepairing(GeyserSession session, GeyserItemStack input, GeyserItemStack material) {
+        Set<String> repairMaterials = input.getMapping(session).getRepairMaterials();
+        return repairMaterials != null && repairMaterials.contains(material.getMapping(session).getJavaIdentifier());
+    }
+
+    private boolean isRenaming(GeyserSession session, AnvilContainer anvilContainer) {
+        if (anvilContainer.getResult().isEmpty()) {
+            return false;
+        }
+        // This should really check the name field, but that requires
+        // the localized name of the item and other pains
+        return !Objects.equals(getCustomName(session, anvilContainer.getInput()), getCustomName(session, anvilContainer.getResult()));
+    }
+
+    private int getTagIntValueOr(GeyserItemStack itemStack, String tagName, int defaultValue) {
+        if (itemStack.getNbt() != null) {
+            Tag tag = itemStack.getNbt().get(tagName);
+            if (tag != null && tag.getValue() instanceof Number value) {
+                return value.intValue();
+            }
+        }
+        return defaultValue;
+    }
+
+    private int getRepairCost(GeyserItemStack itemStack) {
+        return getTagIntValueOr(itemStack, "RepairCost", 0);
+    }
+
+    private boolean hasDurability(GeyserSession session, GeyserItemStack itemStack) {
+        if (itemStack.getMapping(session).getMaxDamage() > 0) {
+            return getTagIntValueOr(itemStack, "Unbreakable", 0) == 0;
+        }
+        return false;
+    }
+
+    private int getDamage(GeyserItemStack itemStack) {
+        return getTagIntValueOr(itemStack, "Damage", 0);
+    }
+
+    private String getCustomName(GeyserSession session, GeyserItemStack itemStack) {
+        if (itemStack.getNbt() != null) {
+            Tag tag = itemStack.getNbt().get("display");
+            if (tag instanceof CompoundTag displayTag) {
+                if (displayTag.get("Name") instanceof StringTag nameTag) {
+                    String name = MessageTranslator.convertToPlainText(nameTag.getValue(), session.getLocale());
```

Comment:
Why do we need to convert to plain text? I'm rather confused about the entire check for a custom name and why using the localized item name is a pain when we have the locale files on hand?

Reply:
In its current state, we probably don't need to convert to plain text, but Minecraft Java ignores formatting when comparing the item's name to the rename field. 
This doesn't seem to be the case on Bedrock though as it allows you to edit formatting of the name, but results in an item with "§" stripped out.

The main pain I had with using the localized name was that some items have different locale key depending on the NBT. This was just a simpler method.
Thinking about it now, there may be an issue with renaming items in other languages as Bedrock items are not guaranteed to have the name as on Java. This may result in items being unintentionally renamed.